src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -597,35 +598,35 @@ internal MemberAccessor MemberAccessorStrategy
         }
 
         /// <summary>
-        /// Whether the options instance has been primed for reflection-based serialization.
+        /// JsonTypeInfo resolver used for serialization.
         /// </summary>
-        internal bool IsInitializedForReflectionSerializer;
+        internal IJsonTypeInfoResolver? _typeInfoResolver;
 
         /// <summary>
         /// Initializes the converters for the reflection-based serializer.
-        /// <seealso cref="InitializeForReflectionSerializer"/> must be checked before calling.
         /// </summary>
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
         [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
-        internal void InitializeForReflectionSerializer()
+        internal void EnsureInitializedForReflectionSerializer()
         {
+            if (_typeInfoResolver != null)
+                return;
+
             RootReflectionSerializerDependencies();
-            Volatile.Write(ref IsInitializedForReflectionSerializer, true);
+            _typeInfoResolver = s_defaultTypeInfoResolver;
             if (_cachingContext != null)
             {
-                _cachingContext.Options.IsInitializedForReflectionSerializer = true;
+                // This only applies if resolver wasn't previously initialized
+                _cachingContext.Options._typeInfoResolver ??= s_defaultTypeInfoResolver;
             }
         }
 
         private JsonTypeInfo GetJsonTypeInfoFromContextOrCreate(Type type)
         {
             JsonTypeInfo? info = _serializerContext?.GetTypeInfo(type);
-            if (info == null && IsInitializedForReflectionSerializer)
+            if (info == null && _typeInfoResolver != null)
```

Comment:
As mentioned, it seems that `serializerContext` and `typeInfoResolver` are competing for the same responsibility. I would recommend:

1. Unifying the two behind a single field. This would require having `JsonSerializerContext` imlpement `IJsonTypeInfo` in this iteration.
2. Factoring the fallback logic out of JsonSerializerOptions (that would also include some of the rooting logic). This might mean moving the sentinel and rooting logics to the default resolver and the JsonSerializerContext implementations of `IJsonTypeInfo`.

Reply:
I've mostly addressed this. The rooting logic is currently assigning FallbackResolver to SerializerContext. This is mostly to preserve current behavior as removing `_serializerContext` would make logic overly complicated for little benefit. Once we make the breaking change (remove fallback logic from context completely) this should simplify